apollo-router-core/src/services/checkpoint/async_checkpoint.rs

```diff
@@ -0,0 +1,301 @@
+use super::Step;
+use futures::future::BoxFuture;
+use std::sync::Arc;
+use tower::{
+    buffer::Buffer, util::BoxCloneService, BoxError, Layer, Service, ServiceBuilder, ServiceExt,
+};
+
+#[allow(clippy::type_complexity)]
+pub struct AsyncCheckpointLayer<S, Request>
+where
+    S: Service<Request> + Send + 'static,
+    Request: Send + 'static,
+    S::Future: Send,
+    S::Response: Send + 'static,
+    <S as Service<Request>>::Error: Into<BoxError> + Send + Sync + 'static,
+{
+    checkpoint_fn: Arc<
+        dyn Fn(
+                Request,
+            ) -> BoxFuture<
+                'static,
+                Result<Step<Request, <S as Service<Request>>::Response>, BoxError>,
+            > + Send
+            + Sync
+            + 'static,
+    >,
+}
+
+#[allow(clippy::type_complexity)]
+impl<S, Request> AsyncCheckpointLayer<S, Request>
+where
+    S: Service<Request> + Send + 'static,
+    Request: Send + 'static,
+    S::Future: Send,
+    S::Response: Send + 'static,
+    <S as Service<Request>>::Error: Into<BoxError> + Send + Sync + 'static,
+{
+    /// Create an `AsyncCheckpointLayer` from a function that takes a Service Request and returns a `Step`
+    pub fn new(
+        checkpoint_fn: impl Fn(
+                Request,
+            ) -> BoxFuture<
+                'static,
+                Result<Step<Request, <S as Service<Request>>::Response>, BoxError>,
+            > + Send
+            + Sync
+            + 'static,
+    ) -> Self {
+        Self {
+            checkpoint_fn: Arc::new(checkpoint_fn),
+        }
+    }
+}
+
+impl<S, Request> Layer<S> for AsyncCheckpointLayer<S, Request>
+where
+    S: Service<Request> + Send + 'static,
+    <S as Service<Request>>::Future: Send,
+    Request: Send + 'static,
+    <S as Service<Request>>::Response: Send + 'static,
+    <S as Service<Request>>::Error: Into<BoxError> + Send + Sync + 'static,
+{
+    type Service = AsyncCheckpointService<
+        BoxCloneService<Request, <S as Service<Request>>::Response, BoxError>,
+        Request,
+    >;
+
+    fn layer(&self, service: S) -> Self::Service {
+        let inner = Buffer::new(service, 20_000);
```

Comment:
Suggest leaving buffer out for now and requiring a clone service.
We can then put the buffer call in ServiceBuilderExt and handle all the other similar layers in the same way.

Reply:
fixed in 7b2b0159